### DISCUSSION TITLE
No Input for Behringer UMC204HD in Jamulus 3.10.0 running on Ubuntu 23.10, Jack displays no errors

### DISCUSSION THREAD

**Discussion by dubnemo** [jamulussoftware/jamulus#3214] [Technical Support & Help] [answered]
I originally thought this was an ALSA 32bit issue that originally was[ causing concerns in Jack](https://github.com/jackaudio/jack2/issues/959), but I have the latest Jack.

I am connected to my server.
Meters display no sound.
People cannot hear me.
Also numerous other applications like OBS and Audacity are working fine via ALSA.
#2125 is Mac-related -- close but unrelated.

**Comment by jujudusud** [comment]
If you use Ubuntu 23.10, you should use PipeWire instead of PulseAudio/JACK. Is it the case?

**Reply by dubnemo** [reply]
@jujudusud 
I don't see an option for PipeWire in the Jamulus -> Settings -> Audio/Network Setup -> Audio droplist.   Only Jack is in the list.
Is that configurable through a back door configuration file (aka environment variables)?

Years ago I recommended direct support for ALSA, then someone hijacked the Issue for portaudio support:
https://github.com/jamulussoftware/jamulus/issues/116

**Reply by jujudusud** [reply]
PipeWire using its pipewire-jack library and replacement server is able to be used with the "JACK" audio setup setting.

Would you please answer my simple question?

**Reply by dubnemo** [reply]
Ouch.  
First, I was not aware that bridge was even out there.   Good to know thank you. 

When Jamulus is running,  Jack is binding directly with ALSA, which should be the lowest overhead without direct support for ALSA in Jamulus.
```
$ ps -ef | grep jack
dubnemo     3304    1198  0 Dec06 ?        00:00:13 /usr/bin/jackdbus auto
dubnemo    20105    1198  2 14:56 ?        00:01:34 /usr/bin/jackd -T -ndefault -dalsa -r48000 -p128 -n2 -D -Chw:Device -Phw:PCH,0 -i1 -I1
dubnemo    20511   20377  0 15:52 pts/0    00:00:00 grep --color=auto jack
```

With or without Jamulus running, PipeWire is running, and the PipeWire to Pulse bridge.  Not the pipewire to Jack process.  pulse is not running, so I could kill 1212.
```
$ ps -ef | grep pipe
dubnemo     1205    1198  0 Dec06 ?        00:00:00 /usr/bin/pipewire
dubnemo     1206    1198  0 Dec06 ?        00:00:00 /usr/bin/pipewire -c filter-chain.conf
dubnemo     1212    1198  0 Dec06 ?        00:00:00 /usr/bin/pipewire-pulse
dubnemo    20529   20377  0 15:56 pts/0    00:00:00 grep --color=auto pipe
```

Second, why would I want to add more layers between the ALSA and Jamulus?  
I/O -> ALSA -> PipeWire -> Jack -> Jamulus.
I appreciate the thought, but PipeWire still uses ALSA/dmix to access the interface, but I see that PipeWire could support Bluetooth headphones (don't have/ need).  

```
JackCtrl logs:
Wed Dec 20 16:12:50 2023: Starting jack server...
Wed Dec 20 16:12:50 2023: JACK server starting in non-realtime mode
Wed Dec 20 16:12:50 2023: self-connect-mode is "Don't restrict self connect requests"
Wed Dec 20 16:12:50 2023: Acquired audio card Audio2
Wed Dec 20 16:12:50 2023: creating alsa driver ... hw:U192k|hw:U192k|128|3|48000|2|0|nomon|hwmeter|-|32bit
Wed Dec 20 16:12:50 2023: configuring for 48000Hz, period = 128 frames (2.7 ms), buffer = 3 periods
Wed Dec 20 16:12:50 2023: ALSA: final selected sample format for capture: 32bit integer little-endian
Wed Dec 20 16:12:50 2023: ALSA: use 3 periods for capture
Wed Dec 20 16:12:50 2023: ALSA: final selected sample format for playback: 32bit integer little-endian
Wed Dec 20 16:12:50 2023: ALSA: use 3 periods for playback
Wed Dec 20 16:12:50 2023: graph reorder: new port 'system:capture_1'
Wed Dec 20 16:12:50 2023: New client 'system' with PID 0
Wed Dec 20 16:12:50 2023: graph reorder: new port 'system:capture_2'
Wed Dec 20 16:12:50 2023: graph reorder: new port 'system:playback_1'
Wed Dec 20 16:12:50 2023: graph reorder: new port 'system:playback_2'
Wed Dec 20 16:12:50 2023: graph reorder: new port 'system:playback_3'
Wed Dec 20 16:12:50 2023: graph reorder: new port 'system:playback_4'
Wed Dec 20 16:12:51 2023: Saving settings to "/home/dubnemo/.config/jack/conf.xml" ...
```

Maybe I am missing something.

**Reply by jujudusud** [reply]
`Wed Dec 20 16:12:50 2023: JACK server starting in non-realtime mode`
This is certainly your problem. Start the JACK server in realtime mode. 
Use "-R" option for the JACK server.

**Reply by dubnemo** [reply]
It previously barked in realtime mode with the same ERROR as below,  but at least Jack started this time.  Previously JackCtrl couldn't start the server at all.

But still no audio while connected to my Jamulus server.  

```
Wed Dec 20 17:03:35 2023: Starting jack server...
Wed Dec 20 17:03:35 2023: JACK server starting in realtime mode with priority 10
Wed Dec 20 17:03:35 2023: self-connect-mode is "Don't restrict self connect requests"
Wed Dec 20 17:03:35 2023: Acquired audio card Audio2
Wed Dec 20 17:03:35 2023: creating alsa driver ... hw:U192k|hw:U192k|128|3|48000|2|0|nomon|hwmeter|-|32bit
Wed Dec 20 17:03:35 2023: configuring for 48000Hz, period = 128 frames (2.7 ms), buffer = 3 periods
Wed Dec 20 17:03:35 2023: ALSA: final selected sample format for capture: 32bit integer little-endian
Wed Dec 20 17:03:35 2023: ALSA: use 3 periods for capture
Wed Dec 20 17:03:35 2023: ALSA: final selected sample format for playback: 32bit integer little-endian
Wed Dec 20 17:03:35 2023: ALSA: use 3 periods for playback
Wed Dec 20 17:03:35 2023: ERROR: Cannot use real-time scheduling (RR/10) (1: Operation not permitted)
Wed Dec 20 17:03:35 2023: ERROR: AcquireSelfRealTime error
Wed Dec 20 17:03:35 2023: graph reorder: new port 'system:capture_1'
Wed Dec 20 17:03:35 2023: New client 'system' with PID 0
Wed Dec 20 17:03:35 2023: graph reorder: new port 'system:capture_2'
Wed Dec 20 17:03:35 2023: graph reorder: new port 'system:playback_1'
Wed Dec 20 17:03:35 2023: graph reorder: new port 'system:playback_2'
Wed Dec 20 17:03:35 2023: graph reorder: new port 'system:playback_3'
Wed Dec 20 17:03:35 2023: graph reorder: new port 'system:playback_4'
Wed Dec 20 17:03:36 2023: Saving settings to "/home/dubnemo/.config/jack/conf.xml" ...
17:03:37.637 JACK connection change.
17:03:37.638 Statistics reset.
17:03:37.647 Client activated.
17:03:37.648 Patchbay deactivated.
17:03:37.651 JACK connection graph change.
Wed Dec 20 17:03:37 2023: New client 'qjackctl' with PID 21229
```

**Reply by dubnemo** [reply]
There is audio microphone level on the interface to be clear.

**Reply by jujudusud** [reply]
Your system is not configured for realtime scheduling of audio. That's why JACK shows :
`ERROR: Cannot use real-time scheduling (RR/10) (1: Operation not permitted)`
`ERROR: AcquireSelfRealTime error`
If you don't know how to set your system up, you should ask Ubuntu's forums.
Or read this page : [JACK FAQ](https://jackaudio.org/faq/linux_rt_config.html)

My advice is to install the entire PipeWire metapackage which includes pipewire-jack OR add Ubuntu Studio 23.10 which includes all the necessary things.

[Ubuntu Doc in french](https://doc.ubuntu-fr.org/jackd) saying that it is not needed to install jack with the latest versions of Ubuntu...

**Reply by dubnemo** [reply]
I see.  Ubuntu Studio is intriguing, and yes, I am using Ubuntu Desktop official release.    Let me focus on the patching, not a complete reinstall.

I still would prefer direct ALSA support for lower level binding.  Too many layers.

**Reply by jujudusud** [reply]
Do you still want to use the standard Ubuntu desktop?
If so, you will use ALSA + PipeWire + JACK or only ALSA + PipeWire.
PipeWire is able to provide:

1. ALSA + PulseAudio +JACK servers, this is the normal use case,
2. ALSA + PulseAudio and become a JACK client providing sinks.

What do you want to do?

**Reply by dubnemo** [reply]
Second option is where I am going.

**Reply by dubnemo** [reply]
Here is where I am at -- without addressing PipeWire at this time.

**Reviewed this:**
https://jackaudio.org/faq/linux_rt_config.html

Added dubnemo user to the audio group
$ `sudo usermod -aG audio dubnemo`
[sudo] password for dubnemo: 

Created the file called /etc/security/limits.d/audio.conf. 
Added the following two lines to this file:
```
@audio   -  rtprio     95
@audio   -  memlock    unlimited
```

Restarted computer

**QjackCtl now is able to support RT scheduling**

```
Thu Dec 21 12:08:34 2023: Starting jack server...
Thu Dec 21 12:08:34 2023: JACK server starting in realtime mode with priority 10
Thu Dec 21 12:08:34 2023: self-connect-mode is "Don't restrict self connect requests"
Thu Dec 21 12:08:34 2023: Acquired audio card Audio2
Thu Dec 21 12:08:34 2023: creating alsa driver ... hw:U192k|hw:U192k|128|3|48000|2|0|nomon|hwmeter|-|32bit
Thu Dec 21 12:08:34 2023: configuring for 48000Hz, period = 128 frames (2.7 ms), buffer = 3 periods
Thu Dec 21 12:08:34 2023: ALSA: final selected sample format for capture: 32bit integer little-endian
Thu Dec 21 12:08:34 2023: ALSA: use 3 periods for capture
Thu Dec 21 12:08:34 2023: ALSA: final selected sample format for playback: 32bit integer little-endian
Thu Dec 21 12:08:34 2023: ALSA: use 3 periods for playback
Thu Dec 21 12:08:34 2023: graph reorder: new port 'system:capture_1'
Thu Dec 21 12:08:34 2023: New client 'system' with PID 0
Thu Dec 21 12:08:34 2023: graph reorder: new port 'system:capture_2'
Thu Dec 21 12:08:34 2023: graph reorder: new port 'system:playback_1'
Thu Dec 21 12:08:34 2023: graph reorder: new port 'system:playback_2'
Thu Dec 21 12:08:34 2023: graph reorder: new port 'system:playback_3'
Thu Dec 21 12:08:34 2023: graph reorder: new port 'system:playback_4'
Thu Dec 21 12:08:35 2023: Saving settings to "/home/dubnemo/.config/jack/conf.xml" ...
12:08:36.909 JACK connection change.
12:08:36.910 Statistics reset.
12:08:36.916 Client activated.
12:08:36.917 Patchbay deactivated.
12:08:36.924 JACK connection graph change.
Thu Dec 21 12:08:36 2023: New client 'qjackctl' with PID 2571
https://wiki.archlinux.org/title/PipeWire
```

**Jamulus - still no audio input from the UMC204HD**
=

I forgot that I previously had the **Roland RCC-10-USXF** running on this machine.  
```
Thu Dec 21 12:14:42 2023: Starting jack server...
Thu Dec 21 12:14:42 2023: JACK server starting in realtime mode with priority 10
Thu Dec 21 12:14:42 2023: self-connect-mode is "Don't restrict self connect requests"
12:14:42.881 D-BUS: JACK server was started (org.jackaudio.service aka jackdbus).
Cannot connect to server socket err = No such file or directory
Cannot connect to server request channel
jack server is not running or cannot be started
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
JackShmReadWritePtr::~JackShmReadWritePtr - Init not done for -1, skipping unlock
Thu Dec 21 12:14:42 2023: Acquired audio card Audio2
Thu Dec 21 12:14:42 2023: Acquired audio card Audio1
Thu Dec 21 12:14:42 2023: creating alsa driver ... hw:PCH,0|hw:Device,0|128|3|48000|1|0|nomon|hwmeter|-|32bit
Thu Dec 21 12:14:42 2023: configuring for 48000Hz, period = 128 frames (2.7 ms), buffer = 3 periods
Thu Dec 21 12:14:42 2023: ALSA: final selected sample format for capture: 16bit little-endian
Thu Dec 21 12:14:42 2023: ALSA: use 3 periods for capture
Thu Dec 21 12:14:42 2023: ALSA: final selected sample format for playback: 32bit integer little-endian
Thu Dec 21 12:14:42 2023: ALSA: use 3 periods for playback
Thu Dec 21 12:14:42 2023: graph reorder: new port 'system:capture_1'
Thu Dec 21 12:14:42 2023: New client 'system' with PID 0
Thu Dec 21 12:14:42 2023: graph reorder: new port 'system:playback_1'
Thu Dec 21 12:14:42 2023: graph reorder: new port 'system:playback_2'
Thu Dec 21 12:14:42 2023: graph reorder: new port 'system:playback_3'
Thu Dec 21 12:14:42 2023: graph reorder: new port 'system:playback_4'
Thu Dec 21 12:14:43 2023: Saving settings to "/home/dubnemo/.config/jack/conf.xml" ...
12:14:44.916 JACK connection change.
12:14:44.916 Statistics reset.
12:14:44.925 Client activated.
12:14:44.925 Patchbay deactivated.
12:14:44.930 JACK connection graph change.
Thu Dec 21 12:14:44 2023: New client 'qjackctl' with PID 2962
12:15:00.429 JACK connection graph change.
12:15:00.571 JACK connection change.
12:15:00.695 JACK connection graph change.
Thu Dec 21 12:15:00 2023: New client 'Jamulus' with PID 2993
Thu Dec 21 12:15:00 2023: Connecting 'system:capture_1' to 'Jamulus:input left'
Thu Dec 21 12:15:00 2023: Connecting 'Jamulus:output left' to 'system:playback_1'
Thu Dec 21 12:15:00 2023: Connecting 'Jamulus:output right' to 'system:playback_2'
Thu Dec 21 12:15:00 2023: New client 'Jamulus-01' with PID 3009
Thu Dec 21 12:15:00 2023: Connecting 'system:capture_1' to 'Jamulus-01:input left'
Thu Dec 21 12:15:00 2023: Connecting 'Jamulus-01:output left' to 'system:playback_1'
Thu Dec 21 12:15:00 2023: Connecting 'Jamulus-01:output right' to 'system:playback_2'
12:15:00.780 JACK connection change.
12:15:05.309 JACK connection graph change.
12:15:05.401 JACK connection change.
Thu Dec 21 12:15:05 2023: Disconnecting 'system:capture_1' from 'Jamulus-01:input left'
Thu Dec 21 12:15:05 2023: Disconnecting 'Jamulus-01:output left' from 'system:playback_1'
Thu Dec 21 12:15:05 2023: Disconnecting 'Jamulus-01:output right' from 'system:playback_2'
Thu Dec 21 12:15:05 2023: Client 'Jamulus-01' with PID 3009 is out
12:15:17.249 JACK connection graph change.
12:15:17.430 JACK connection change.
Thu Dec 21 12:15:17 2023: Disconnecting 'system:capture_1' from 'Jamulus:input left'
Thu Dec 21 12:15:17 2023: Disconnecting 'Jamulus:output left' from 'system:playback_1'
Thu Dec 21 12:15:17 2023: Disconnecting 'Jamulus:output right' from 'system:playback_2'
Thu Dec 21 12:15:17 2023: Client 'Jamulus' with PID 2993 is out
```

The Roland RCC-10-USXF XLR->USB interface works just fine in Jamulus.  
=
This interface is not duplex so I route the output the laptop headphones.  

But the interface I want to use I want to use if the UMC204HD, since the Roland input level is too low and I had to use a pre-amp and the pre-amp gave out recently.

So that seems to suggest that PipeWire is not required and my machine is setup OK.

**Reply by dubnemo** [reply]
Capturing the latency without PipeWire, before I configure PipeWire
![image](https://github.com/jamulussoftware/jamulus/assets/32434235/3b407931-17d7-49f5-b79b-03decc094deb)

**Reply by jujudusud** [reply]
Ok. In order to use JACK in your case (Ubuntu 23.10), you have to follow this guide :
https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/Config-JACK#jack-bridge

But, it is not the default use case.

**Reply by dubnemo** [reply]
I am back at this issue (post-holiday) and I now have pipewire install, running but clearly not configured as I am now getting feedback in Jamulus.  

![image](https://github.com/jamulussoftware/jamulus/assets/32434235/7efcc442-5590-411f-a616-ac41bd6a84e5)


It seems underneath the hood somewhere, there is a lookback from output to input.     iWhile this is a step-forward, it's not the solution.   

dubnemo@dubnemo-20BX001PUS:~$ ps -ef | grep pipewire
dubnemo     1533    1485  0 Jan10 ?        00:00:02 /usr/bin/pipewire
dubnemo     1534    1485  0 Jan10 ?        00:00:00 /usr/bin/pipewire -c filter-chain.conf
dubnemo     1535    1485  0 Jan10 ?        00:00:00 /usr/bin/pipewire-pulse
dubnemo     5847    5837  0 13:55 pts/0    00:00:00 grep --color=auto pipewire

The Jack bridge link above states:
Since 0.3.71, The recommended way to start PipeWire as a JACK client is to use jackdbus and the [module-jackdbus-detect](https://docs.pipewire.org/page_module_jackdbus_detect.html).
You can add this module to the PipeWire main config file for highest performance. It is also possible to load the module into the pipewire-pulse server with pactl load-module module-jackdbus-detect.
Since 0.3.81, the jackdbus module is already loaded and ready to use.

I am not clear what library "version 0.3.81" is referring to.  PipeWire seems to be: 
dubnemo@dubnemo-20BX001PUS:~$ pipewire --version
pipewire
Compiled with libpipewire 1.0.0
Linked with libpipewire 1.0.0

https://docs.pipewire.org/page_module_jackdbus_detect.html  is not helpful, as I see there are numerous configuration files, and the documentation fails to clarify which file this configuration should be placed.

**Reply by jujudusud** [reply]
The best way is to use :
`https://ubuntustudio.org/ubuntu-studio-installer/#audioconfig`
All the config will be done flawlessly.

Replace JACK by pipewire-jack, PulseAudio by pipewire-pulse and ALSA by pipewire-alsa.

`https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/Config-JACK#jack-server`

**Reply by dubnemo** [reply]
I use this computer for VS Code, amongst other things, so a rebuild is not happening anytime soon.
Pipewire seems too over-engineered for a simple use case where ALSA alone would suffice.
Less is more.

**Reply by jujudusud** [reply]
You need to code Jamulus to use ALSA instead of JACK as I already told you or use native/PipeWire replacement JACK server.
For now, it is JACK only.
If less is more a goal, remove PulseAudio and PipeWire completly from your system and stay with only ALSA + native JACK server.

**Reply by dubnemo** [reply, marked as answer]
Update:  
After complete uninstall of pipewire and pulseaudio, Jack/ALSA approach worked.  I do have to manually start the Jack server via qjackctl, but that is minor as I am hearing and seeing audio.

One additional comment: the uninstall blew away my Ubuntu Deskop, and that had to be reinstalled.

It has been ~35 years since I have developed in C code.  Is there a getting started wiki somewhere describing expected interfaces and other developer guidelines?  But there are a ton of repos out there with existing ALSA code.

**Reply by pcar75** [reply]
Note that Jamulus is written in C++ 11 and the Qt and Opus interfaces/libraries/frameworks are almost languages of their own.
Refs.:
https://github.com/jamulussoftware/jamulus
https://github.com/jamulussoftware/jamulus/blob/main/CONTRIBUTING.md
https://en.cppreference.com/w/cpp
and of course Bjarne Stroustrup's https://www.stroustrup.com/index.html
- "Programming: Principles and Practice using C++ (2nd edition)" 1312 pages
- "The C++ Programming Language (4th edition)" 1376 pages ... if pursuing a doctorate ;-|

Good Luck !

**Reply by dubnemo** [reply]
I have the 2nd edition of The C++ Programming Language, only 699 pages.  
Probably my least favorite language.  :-(
